proof/refine/StateRelation.thy

```diff
@@ -0,0 +1,444 @@
+  where
+  "thread_state_relation (Structures_A.Running) ts'
+     = (ts' = Structures_H.Running)"
+| "thread_state_relation (Structures_A.Restart) ts'
+     = (ts' = Structures_H.Restart)"
+| "thread_state_relation (Structures_A.Inactive) ts'
+     = (ts' = Structures_H.Inactive)"
+| "thread_state_relation (Structures_A.IdleThreadState) ts'
+     = (ts' = Structures_H.IdleThreadState)"
+| "thread_state_relation (Structures_A.BlockedOnReply) ts'
+     = (ts' = Structures_H.BlockedOnReply)"
+| "thread_state_relation (Structures_A.BlockedOnReceive oref sp) ts'
+     = (ts' = Structures_H.BlockedOnReceive oref (receiver_can_grant sp))"
+| "thread_state_relation (Structures_A.BlockedOnSend oref sp) ts'
+     = (ts' = Structures_H.BlockedOnSend oref (sender_badge sp)
+                         (sender_can_grant sp) (sender_can_grant_reply sp) (sender_is_call sp))"
+| "thread_state_relation (Structures_A.BlockedOnNotification oref) ts'
+     = (ts' = Structures_H.BlockedOnNotification oref)"
+
+definition tcb_relation :: "Structures_A.tcb \<Rightarrow> Structures_H.tcb \<Rightarrow> bool" where
+  "tcb_relation \<equiv> \<lambda>tcb tcb'.
+     tcb_fault_handler tcb = to_bl (tcbFaultHandler tcb')
+   \<and> tcb_ipc_buffer tcb = tcbIPCBuffer tcb'
+   \<and> arch_tcb_relation (tcb_arch tcb) (tcbArch tcb')
+   \<and> thread_state_relation (tcb_state tcb) (tcbState tcb')
+   \<and> fault_rel_optionation (tcb_fault tcb) (tcbFault tcb')
+   \<and> cap_relation (tcb_ctable tcb) (cteCap (tcbCTable tcb'))
+   \<and> cap_relation (tcb_vtable tcb) (cteCap (tcbVTable tcb'))
+   \<and> cap_relation (tcb_reply tcb) (cteCap (tcbReply tcb'))
+   \<and> cap_relation (tcb_caller tcb) (cteCap (tcbCaller tcb'))
+   \<and> cap_relation (tcb_ipcframe tcb) (cteCap (tcbIPCBufferFrame tcb'))
+   \<and> tcb_bound_notification tcb = tcbBoundNotification tcb'
+   \<and> tcb_mcpriority tcb = tcbMCP tcb'
+   \<and> tcb_priority tcb = tcbPriority tcb'
+   \<and> tcb_time_slice tcb = tcbTimeSlice tcb'
+   \<and> tcb_domain tcb = tcbDomain tcb'"
+
+\<comment> \<open>
+  A pair of objects @{term "(obj, obj')"} should satisfy the following relation when, under further
+  mild assumptions, a @{term corres_underlying} lemma for @{term "set_object obj"}
+  and @{term "setObject obj'"} can be stated: see setObject_other_corres in KHeap_R.
+
+  TCBs do not satisfy this relation because the tcbSchedPrev and tcbSchedNext fields of a TCB are
+  used to model the ready queues, and so an update to such a field would correspond to an update
+  to a ready queue (see ready_queues_relation below).\<close>
+definition
+  other_obj_relation :: "Structures_A.kernel_object \<Rightarrow> Structures_H.kernel_object \<Rightarrow> bool"
+where
+  "other_obj_relation obj obj' \<equiv>
+   (case (obj, obj') of
+      (Endpoint ep, KOEndpoint ep') \<Rightarrow> ep_relation ep ep'
+    | (Notification ntfn, KONotification ntfn') \<Rightarrow> ntfn_relation ntfn ntfn'
+    | _ \<Rightarrow> False)"
```

Comment:
Was probably there before, but the outer parentheses around `case` are not needed (only for `=`, not for `==`)

Reply:
done